modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/http/conn/ProxyConfig.java

```diff
@@ -85,6 +116,196 @@ public HttpHost selectProxy(final HttpHost target) {
+     *
+     * If the request hits the ESB for the first time, checks whether the default profile is configured. If so, a flag
+     * is set true. Then the targetHost is matched against the proxyProfileMaps key set
+     * i.e check any of the key patten is matching with the targetHost. If the targetHost is matched against a key then
+     * calls getProxyProfileConfig(String, String) and returns the proxyProfileConfig.
+     *
+     * If the targetHost is not matched against the proxyProfileMap key set and default profile flag is set true
+     * then calls getProxyProfileConfig(String, String) and returns the defaultProfile.
+     *
+     * @param targetHost request's targeted host
+     * @return ProxyProfileConfig for the given targetHost
+     */
+    private ProxyProfileConfig getProxyProfileForTargetHost(String targetHost) {
+
+        if (knownProxyConfigMap.containsKey(targetHost)) {
+            return knownProxyConfigMap.get(targetHost);
+        }
+
+        if (knownDirectHosts.contains(targetHost)) {
+            return null;
+        }
+
+        boolean defaultProfile = false;
+        for (String key : proxyProfileMap.keySet()) {
+            if (key.equals("*")) {
+                defaultProfile = true;
+                continue;
+            }
+            if (targetHost.matches(key)) {
+                return getProxyProfileConfig(targetHost, key);
+            }
+        }
+
+        if (defaultProfile) {
+            return getProxyProfileConfig(targetHost, "*");
+        }
+
+        return null;
+    }
+
+    /**
+     * Selects the proxyProfile for the given key and gets the bypass set. Matches the targetHost against the
+     * bypass set. If it is matched then adds the targetHost to the knownDirectHosts List and returns null.
+     * Otherwise (i.e the targetHost is not matched in the bypass proxy) puts the proxyProfile against the targetHost
+     * into the knownProxyConfigMap and returns the proxyProfileConfig
+     *
+     * @param targetHost request's targeted host
+     * @param key proxyProfileMap's key, if default profile then the key is "*"
+     * @return proxyProfileConfig
+     */
+    private ProxyProfileConfig getProxyProfileConfig(String targetHost, String key) {
+        ProxyProfileConfig proxyProfileConfig = proxyProfileMap.get(key);
+        Set<String> proxyByPass = proxyProfileConfig.getProxyByPass();
+        for (String bypass : proxyByPass) {
+            if (targetHost.matches(bypass)) {
+                knownDirectHosts.add(targetHost);
+                return null;
+            }
+        }
+        knownProxyConfigMap.put(targetHost, proxyProfileConfig);
+        return proxyProfileConfig;
+    }
+
+
+    /**
+     * select the proxy credential for the targetHost
+     *
+     * @param targetHost targeted host
+     * @return proxy credential for the given end point, if not returns null
+     */
+    public UsernamePasswordCredentials getCredentialsForTargetHost(String targetHost) {
+
```

Comment:
Removed unnecessary space. Also I will check the other locations and remove them
Thanks